math/src/main/kotlin/ktx/math/ImmutableVector2.kt

```diff
@@ -0,0 +1,218 @@
+
+    override val len2: Float = Vector2.len2(x, y)
+
+    override val nor: ImmutableVector2 get() = withLength2(1f)
+
+    override fun isZero(margin: Float): Boolean = (x == 0f && y == 0f) || len2 < margin
+
+    override fun unaryMinus(): ImmutableVector2 = ImmutableVector2(-x, -y)
+
+    override operator fun minus(other: ImmutableVector2): ImmutableVector2 = minus(other.x, other.y)
+
+    /** Returns the result of subtracting the given vector from this vector */
+    fun minus(deltaX: Float = 0f, deltaY: Float = 0f): ImmutableVector2 = ImmutableVector2(x - deltaX, y - deltaY)
+
+    override operator fun plus(other: ImmutableVector2): ImmutableVector2 = plus(other.x, other.y)
+
+    /** Returns the result of adding the given vector to this vector */
+    fun plus(deltaX: Float = 0f, deltaY: Float = 0f): ImmutableVector2 = ImmutableVector2(x + deltaX, y + deltaY)
+
+    override fun inc(): ImmutableVector2 = ImmutableVector2(x + 1, y + 1)
+    override fun dec(): ImmutableVector2 = ImmutableVector2(x - 1, y - 1)
+
+    override operator fun times(scalar: Float): ImmutableVector2 = times(scalar, scalar)
+    override operator fun times(vector: ImmutableVector2): ImmutableVector2 = times(vector.x, vector.y)
+
+    /** Returns a new vector instance scaled by the given [factorX] and [factorY] factors */
+    fun times(factorX: Float, factorY: Float): ImmutableVector2 = ImmutableVector2(x * factorX, y * factorY)
+
+    override fun withLimit2(limit2: Float): ImmutableVector2 =
+            if (len2 <= limit2) this else withLength2(limit2)
+
+    override fun withClamp2(min2: Float, max2: Float): ImmutableVector2 {
+        val l2 = len2
+
+        return when {
+            l2 < min2 -> withLength2(min2)
+            l2 > max2 -> withLength2(max2)
+            else -> this
+        }
+    }
+
+    override fun withLength2(length2: Float): ImmutableVector2 {
+        val oldLen2 = len2
+
+        return if (oldLen2 == 0f || oldLen2 == length2) this else times(sqrt(length2 / oldLen2))
+    }
+
+    override fun dot(vector: ImmutableVector2): Float = dot(vector.x, vector.y)
+
+    /** Returns the dot product of this vector by the given vector */
+    fun dot(otherX: Float, otherY: Float): Float = Vector2.dot(x, y, otherX, otherY)
+
+    override fun dst2(vector: ImmutableVector2): Float = dst2(vector.x, vector.y)
+
+    /**
+     * Returns the squared distance between this and the other vector
+     *
+     * This method is faster than [dst] because it avoids calculating a square root. It is useful for comparisons,
+     * but not for getting exact distance, as the return value is the square of the actual distance.
+     */
+    fun dst2(otherX: Float, otherY: Float): Float = Vector2.dst2(x, y, otherX, otherY)
+
+    /** @return the distance between this and the other vector */
+    fun dst(otherX: Float, otherY: Float): Float = Vector2.dst(x, y, otherX, otherY)
+
+    /** Returns the result of multiplying this vector by the given matrix */
```

Comment:
I'd include a comment about the assumptions that this method makes. You cannot normally multiply a 2D vectors with a 3D matrix, can you?

Reply:
is `apply the given affine transformation and return the resulting vector` better`? It would make  clearer that we assume that the matrix is an `affine transformation` representation.

Reply:
I meant that this method assumes the vector has a `z=1` component during the multiplication (hence the addition of unmodified `val[6]` and `val[7]`) and "removes" the last element after the operation. This might not be obvious, especially for someone who does not deal with graphics very often.

Reply:
>this method assumes the vector has a z=1 component

Yes, because it assumes it is an *affine transformation*.

>This might not be obvious, especially for someone who does not deal with graphics very often. 

Yes. But I'm a bit affraid of making it look complicated if we talk about `z=1`, wherever if we only took about *affine transformation*, there is no need for the user to understand how they work. he only need to pass a valid affine transformation, which he can get from LibGDX.

Actually, it makes me think we may change the signature for `operator times(transform: Affine2): Vector2`. It would make it clearer about what the contract is. And experienced user can easily create an `Affine2` out of a `Matrix3`.

Reply:
And I may also rewrite the test for something more meaningful.

For instance:

```kotlin
val vector = ImmutableVector(3f, 4f)

transformation = Affine2()
transformation.preTranslate(-2f, 3f)
transformation.preRotate(18f)
transformation.preScale(1.5f, 2f)

assertEquals(vector.plus(-2f, 3f).withRotation(18f).scl(1.5f, 2f), vector * transformation)
```

Reply:
> Yes, because it is how *affine transformation* matrices are defined.

Which isn't mentioned in the documentation or the method name, hence the comment. Method name suggests a simple matrix multiplication, in which case the dimensions simply would not match. I think mentioning both *how* and *why* is worth including in the doc.

Reply:
>Which isn't mentioned in the documentation or the method name, hence the comment. 

Yes, of course, it should be added. I'm not challenging your initial comment. I also think it deserve to be clarified.

To be clear, I propose to change the documentation and signature to the following:

```kotlin
/** Apply the given affine [transformation] and return the resulting vector */
operator fun times(transformation: Affine2): ImmutableVector2
``` 

Here method signature and comment would clearly state that an affine transformation is needed and that the function returns the result of applying the transformation to the vector.

Whould that be ok for you?

Reply:
I like this approach. The original LibGDX `Vector2` has the misleading `mul(Matrix3)`, so it's a matter of compatibility vs usability vs clarity of API. I think we can add a deprecated `mul(Matrix3)` method (since the original mutates) that points to `times(Affine2)` - `Affine2` is fairly easy to obtain with a `Matrix3` either way.